File: Data Structures & Algorithms/generate-parentheses/submission-0.java
class Solution {
    List<String> res;
    int n;

    public List<String> generateParenthesis(int n) {
        res = new ArrayList<>();
        this.n = n;
        StringBuilder sb = new StringBuilder();

        backtrack(0, 0, sb);

        return res;
    }

    private void backtrack(int open, int close, StringBuilder sb) {
        if (sb.length() == n * 2) {
            res.add(sb.toString());
            return;
        }

        if (open < n) {
            sb.append("(");
            backtrack(open + 1, close, sb);
            sb.delete(sb.length() - 1, sb.length());
        }

        if (close < n && open > close) {
            sb.append(")");
            backtrack(open, close + 1, sb);
            sb.delete(sb.length() - 1, sb.length());
        }
    }

    /*
        Each discrete choice (n * 2 total):
            Add ( (push to stack)
            Add ) (pop from stack)
    */
}

File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        memo = []
        n = len(s)

        for i in range(n):
            memo.append([0] * n)

        for i in range(n):
            l, r = i, i
            while l >= 0 and r < n and s[l] == s[r]:
                memo[l][r] = 1
                l -= 1
                r += 1

            l, r = i, i + 1
            while l >= 0 and r < n and s[l] == s[r]:
                memo[l][r] = 1
                l -= 1
                r += 1
        
        res = 0

        for i in memo:
            for j in i:
                res += j
        
        return res
